File: leetcode/0337-house-robber-iii/0337-house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * House Robber III - Maximum money that can be robbed from binary tree houses
     * where directly connected houses cannot be robbed together
     * 
     * @param root The root of the binary tree
     * @return Maximum amount of money that can be robbed
     */
    public int rob(TreeNode root) {
        // Get the maximum values for both scenarios (rob root vs don't rob root)
        int[] result = dfs(root);
      
        // Return the maximum between robbing root (result[0]) and not robbing root (result[1])
        return Math.max(result[0], result[1]);
    }
  
    /**
     * DFS helper method that returns optimal values for current node
     * 
     * @param root Current node being processed
     * @return Array where [0] = max money if current node is robbed,
     *                     [1] = max money if current node is not robbed
     */
    private int[] dfs(TreeNode root) {
        // Base case: null node contributes 0 to both scenarios
        if (root == null) {
            return new int[2];
        }
      
        // Recursively calculate optimal values for left and right subtrees
        int[] leftSubtree = dfs(root.left);
        int[] rightSubtree = dfs(root.right);
      
        // Calculate optimal values for current node
        // robCurrent: If we rob current node, we cannot rob its children
        int robCurrent = root.val + leftSubtree[1] + rightSubtree[1];
      
        // skipCurrent: If we skip current node, we can choose the max from each child
        int skipCurrent = Math.max(leftSubtree[0], leftSubtree[1]) + 
                         Math.max(rightSubtree[0], rightSubtree[1]);
      
        return new int[] {robCurrent, skipCurrent};
    }
}

File: 0337-house-robber-iii/0337-house-robber-iii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * House Robber III - Rob houses arranged in a binary tree without robbing adjacent nodes
 * @param root - Root node of the binary tree
 * @returns Maximum amount that can be robbed
 */
function rob(root: TreeNode | null): number {
    /**
     * DFS helper function that returns two values for each node:
     * [0] - Maximum amount if current node is robbed
     * [1] - Maximum amount if current node is not robbed
     */
    const dfs = (node: TreeNode | null): [number, number] => {
        // Base case: empty node contributes nothing
        if (!node) {
            return [0, 0];
        }
      
        // Recursively process left and right subtrees
        const [leftRobbed, leftNotRobbed] = dfs(node.left);
        const [rightRobbed, rightNotRobbed] = dfs(node.right);
      
        // If we rob current node, we cannot rob its children
        const robCurrent = node.val + leftNotRobbed + rightNotRobbed;
      
        // If we don't rob current node, we can choose to rob or not rob its children
        const skipCurrent = Math.max(leftRobbed, leftNotRobbed) + Math.max(rightRobbed, rightNotRobbed);
      
        return [robCurrent, skipCurrent];
    };
  
    // Return the maximum of robbing or not robbing the root
    const [robRoot, skipRoot] = dfs(root);
    return Math.max(robRoot, skipRoot);
}
